### PR TITLE
Strip whitespace for proxy urls, test for same. #4613

### PR DESCRIPTION
**Pull request by ghost**
I'm not sure this is the right place to do this operation, but it does solve the problem. Or at least it passes the test that I wrote for it.

Quite likely I'm not reading them correctly, but as far as I can tell the other test methods related to proxies don't test something like a `requests.get()` call with the expectation of a successful result.

I found a ton of tests for error cases, though, and modeled the test for this situation on those, but it seems like there would be a better way.

### CODE REVIEW

File: requests/adapters.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def get_connection(self, url, proxies=None):
         proxy = select_proxy(url, proxies)
 
         if proxy:
-            proxy = prepend_scheme_if_needed(proxy, 'http')
+            proxy = prepend_scheme_if_needed(proxy, 'http').strip()
```
Comment (nateprewitt) [inline requests/adapters.py:303]: We may want to call `strip` on the proxy itself before handing it to `prepend_scheme_if_needed`. Leading whitespace causes issues with `urlparse` and we get back a weird url which will fail later.

File: tests/test_requests.py [hunk 2]
```diff
@@ -549,6 +549,11 @@ def test_proxy_error_on_bad_url(self, httpbin, httpbin_secure):
         with pytest.raises(InvalidProxyURL):
             requests.get(httpbin(), proxies={'http': 'http:///example.com:8080'})
 
+    def test_proxy_strip_spaces(self):
+        # Don't yet have testing proxies sorted out.
+        with pytest.raises(ProxyError):
+            requests.get('http://localhost:1', proxies={'http': 'http://localhost:1234 '})
```
Comment (nateprewitt) [inline tests/test_requests.py]: Would you mind adding a couple more tests for proxies without schemes, leading whitespace, and proxy values of `None`? You should be able to find examples of using pytest's `parametrize` decorator above to do all the tests in one function.

File: requests/adapters.py (hunk 1 above)
Comment (sigmavirus24) [reply requests/adapters.py:303]: I'm not even certain either of those are the right change, honestly, However, I can't think of a better change.

Comment (nateprewitt) [reply requests/adapters.py:303]: I agree this isn’t great, but I’m not sure if we have a better place to catch both environmental and parameter proxies for sanitizing. We’d need to make a larger change to how proxies are handled I think if we wanted this to be done in a less hacky fashion.

Comment (nateprewitt) [reply requests/adapters.py:303]: Sorry, let me redact that last statement. This doesn’t cover `request_url`, so we really need to do a single mutation on the proxies dict passed to send, or ensure the malformed uri exception is raised for the trailing white space.

Comment (ghost) [reply requests/adapters.py:303]: Do you mean to do something like this?
```python
if proxy:
    proxy.strip()
```

File: tests/test_requests.py (hunk 2 above)
Comment (ghost) [reply tests/test_requests.py]: - [ ] missing scheme
- [ ] leading whitespace
- [ ] proxy value None

**Comment by sethmlarson** [comment]
In an effort to clean up the issue tracker to only have issues that are still relevant to the project we've done a quick pass and decided this issue may no longer be relevant for a variety of potential reasons, including:

* Applies to a much older version, unclear whether the issue still applies.
* Change requires a backwards incompatible release and it's unclear if the benefits are worth the migration effort from the community.
* There isn't a clear demand from the community on the change landing in Requests.

If you think the issue should remain open, please comment so below or open a new issue and link back to the original issue. Again, thank you for opening the issue and for the discussion, it's much appreciated.
